src/routes/files.ts

```diff
@@ -1,14 +1,351 @@
+import { asEither, asMap, asNull, asObject } from 'cleaners'
 import Router from 'express-promise-router'
 import { dataStore } from '../db'
 import {
+  getFilenameFromPath,
+  getParentPathsOfPath,
+  mergeDirectoryFilePointers,
+  normalizePath,
+  updateDirectoryFilePointers
+} from '../utils'
+import {
+  ApiClientError,
   ApiErrorResponse,
   ApiResponse,
+  asStoreDirectoryDocument,
+  asStoreFile,
+  asStoreFileDocument,
+  asStoreRootDocument,
   DocumentRequest,
-  Results
+  Results,
+  StoreDirectory,
+  StoreDirectoryDocument,
+  StoreFile,
+  StoreFileDocument,
+  StoreRootDocument
 } from '../types'
 
+type FilesPostBody = ReturnType<typeof asFilesPostBody>
+
+interface FilesPostResponseData {
+  paths: {
+    [path: string]: number
+  }
+}
+
+type FileKeyInfoMap = Map<string, { repoid: string; path: string }>
+
+const asFilesPostBody = asObject({
+  paths: asMap(asEither(asStoreFile, asNull))
+})
+
 export const filesRouter = Router()
 
+// ---------------------------------------------------------------------
+// POST /files
+// ---------------------------------------------------------------------
+
+filesRouter.post('/files', async (req, res, next) => {
+  let body: FilesPostBody
+  let fileKeys: string[]
+  const fileKeyInfoMap: FileKeyInfoMap = new Map()
+
+  // Validate request body
+  try {
+    body = asFilesPostBody(req.body)
+
+    // Validate fileKeys are formated correctly (includes repoid)
+    fileKeys = Object.keys(body.paths)
+    fileKeys.forEach(fileKey => {
+      const [repoid, ...pathPart] = fileKey.split(':')
+      const path = normalizePath(pathPart.join(':'))
+
+      if (!repoid) {
+        throw new Error(`Missing repoid for key: '${fileKey}'`)
+      }
+      if (!path) {
+        throw new Error(`Missing path for key: '${fileKey}'`)
+      }
+      if (path === '/') {
+        throw new Error(
+          `Invalid path '${path}' for key '${fileKey}'.` +
+            `Path cannot be root.`
+        )
+      }
+
+      fileKeyInfoMap.set(fileKey, { repoid, path })
```

Comment:
Is there a reason to use Map for this object? Propose to good ol' object.

Reply:
Furthermore, reduce this code down to validating paths without normalization. If `fileKeys` are invalid, throw an error rather than clean. Why: Because there would be a mismatch between what the client sent and to what the server resolves.